### PR TITLE
1 feature parse tracab data

### LINKED ISSUE #1: Parse Tracab tracking data
Make a function in `databall/tracking_data/parse_tracking_data.py` that reads in the `.dat` tracking data and returns the data. Maybe add a normalize direction function alread as basic preprocessing?

### PR DESCRIPTION
**Pull request by DaanGro**
resolve #1

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -16,6 +16,8 @@ pandas = "^1.5.3"
 numpy = "^1.24.1"
 beautifulsoup4 = "^4.11.1"
 lxml = "^4.9.2"
+tqdm = "4.64.1"
+matplotlib = "^3.6.3"
```
Comment (Alek050) [inline pyproject.toml:20]: I think matplotlib should be removed, is not part of this pull request right?

File: databallpy/load_data/tracking_data/tracab.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+import bs4
+import numpy as np
+import pandas as pd
+from bs4 import BeautifulSoup
+from tqdm import tqdm
+
+from databallpy.load_data.metadata import Metadata
+
+
+def load_tracking_data_tracab(
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:12]: For opta it is called `load_opta_event_data`, so maybe call this one `load_tracab_tracking_data` for consistancy?

File: databallpy/load_data/tracking_data/tracab.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    tracab_loc: str, meta_data_loc: str, verbose: bool = True
+) -> Tuple[pd.DataFrame, Metadata]:
+    """Function to load tracking data and metadata in the tracab format
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:15]: `Function to load tracking data and metadata from the tracab format`

File: databallpy/load_data/tracking_data/tracab.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+    Args:
+        tracab_loc (str): location of the tracking_data.dat file
+        meta_data-loc (str): location of the meta_data.xml file
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:19]: `meta_data_loc`

File: databallpy/load_data/tracking_data/tracab.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        verbose (bool): whether to print info, defaults to True
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:20]: Maybe a bit more elaborate, info about the progress of loading the file in the terminal

File: databallpy/load_data/tracking_data/tracab.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+    Returns:
+        Tuple[pd.DataFrame, Metadata], the tracking data and metadate class
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:23]: metadate --> metadata

File: databallpy/load_data/tracking_data/tracab.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+            data = _add_player_data_to_dict(player, data, idx)
+
+        data = _add_ball_data_to_dict(ball_info, data, idx)
+
+    df = pd.DataFrame(data)
+
+    for col in df.columns:
+        if "_x" in col or "_y" in col or "_z" in col or "_speed" in col:
+            df[col] /= 100
+            df[col] = np.round(df[col], 3)
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:84]: Is this just translating cm to meters? Maybe add a little comment above what your are doing here.

File: databallpy/load_data/tracking_data/tracab.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        else:
+            pass
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:86]: This is unnecissary, just delete it. `else: pass` is the same as just deleting these rows.

File: databallpy/load_data/tracking_data/tracab.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    if f"{team}_{shirt_num}_x" not in data.keys():  # create keys for new player
+        data[f"{team}_{shirt_num}_x"] = [np.nan] * len(data["timestamp"])
+        data[f"{team}_{shirt_num}_y"] = [np.nan] * len(data["timestamp"])
+        data[f"{team}_{shirt_num}_speed"] = [np.nan] * len(data["timestamp"])
+
+    data[f"{team}_{shirt_num}_x"][idx] = int(x)
+    data[f"{team}_{shirt_num}_y"][idx] = int(y)
+    data[f"{team}_{shirt_num}_speed"][idx] = float(speed)
+
+    return data
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:121]: This function could potentially also be used with other provides, if `team_id, _, shirt_num, x, y, speed = player.split(",")` would be left out. I would suggest adding these to the input of this function, where `speed` for instance could be a optional input since maybe not all providers have this variable.

File: databallpy/load_data/tracking_data/tracab.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+    x, y, z, speed, posession, status = ball_info.split(";")[0].split(",")[:6]
+    data["ball_x"][idx] = int(x)
+    data["ball_y"][idx] = int(y)
+    data["ball_z"][idx] = int(z)
+    data["ball_speed"][idx] = float(speed)
+    data["ball_posession"][idx] = posession
+    data["ball_status"][idx] = status
+
+    return data
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:144]: This function does nothing complicated or any calculations/conditions, would just add this to the `_get_tracking_data()` function.

File: databallpy/load_data/tracking_data/tracab.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+    meta_data = Metadata(
+        match_id=match_id,
+        pitch_dimensions=[pitch_size_x, pitch_size_y],
+        match_start_datetime=match_start_datetime,
+        periods_frames=df_frames,
+        frame_rate=frame_rate,
+        home_team_id=home_team_id,
+        home_team_name=home_team_name,
+        home_players=df_home_players,
+        home_score=None,
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:223]: Should be `np.nan`, it is a float, not a string

File: databallpy/load_data/tracking_data/tracab.py [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+    file.close()
+
+    return meta_data
+
+
+def _get_player_data(team: bs4.element.Tag) -> pd.DataFrame:
+    """Function that creates a df containing info on all players for a team
+
+    Args:
+        team (bs4.element.Tag): containing info no all players of a team
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:241]: typo?

File: tests/load_data/event_data/test_tracab.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+                "home_17_x": [1.22, 1.21, 1.21, np.nan, 1.21],
+                "home_17_y": [-13.16, -13.16, -13.17, np.nan, -13.18],
+                "home_17_speed": [0.002, 0.002, 0.002, np.nan, 0.002],
+            }
+        )
+
+    def test_get_metadata(self):
+
+        meta_data = _get_meta_data(self.metadata_loc)
+        assert meta_data == self.expected_metadata
```
Comment (Alek050) [inline tests/load_data/event_data/test_tracab.py:79]: chose `metadata` or `meta_data`, I think I called it `metadata`

File: tests/load_data/event_data/test_tracab.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_get_tracking_data(self):
+        tracking_data = _get_tracking_data(self.tracking_data_loc, verbose=True).fillna(
+            1000
+        )
+        assert tracking_data.equals(self.expected_tracking_data.fillna(1000))
```
Comment (Alek050) [inline tests/load_data/event_data/test_tracab.py:85]: pandas has its own testing moduel. In this case you could do `pd.testing.assert_frame_equal(df1, df2)`, and I would suggest without the `.fillna()` because we also don't use that in the real function.

File: tests/load_data/event_data/test_tracab.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+            self.tracking_data_loc, self.metadata_loc, verbose=True
+        )
+        tracking_data = tracking_data.fillna(1000)
+
+        assert all(
+            [
+                metadata == self.expected_metadata,
+                tracking_data.equals(self.expected_tracking_data.fillna(1000)),
+            ]
+        )
```
Comment (Alek050) [inline tests/load_data/event_data/test_tracab.py:98]: Make two different asserts, this will make it easier to find mistakes and where they originate from if tests fail.

File: tests/load_data/event_data/test_tracab.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
+import unittest
+
+import numpy as np
+import pandas as pd
+
+from databallpy.load_data.metadata import Metadata
+from databallpy.load_data.tracking_data.tracab import (
+    _get_meta_data,
```
Comment (Alek050) [inline tests/load_data/event_data/test_tracab.py:8]: Not all functions have individual tests, this is necesssary, also if you already get 100% covered. If an error emerges it might be hard to say if a specific function fails or if the flow is wrong (for instance the input given to the next function is wrong). It will be a lot easier to find mistakes if all unique functions have their own tests. So also add a test for `_get_player_data`. I think you would make it yourself a lot easier if you input in that function a list of dictionaries for every player instead of the bs4.element.Tag.

File: tests/test_data/tracab_metadata_test.xml [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        <LastName>Blok</LastName>
+        <JerseyNo>1</JerseyNo>
+        <StartFrameCount>100</StartFrameCount>
+        <EndFrameCount>1200</EndFrameCount>
+      </Player>
+      <Player>
+        <PlayerId>450445</PlayerId>
+        <FirstName>Test</FirstName>
+        <LastName>Speler</LastName>
+        <JerseyNo>2</JerseyNo>
```
Comment (Alek050) [inline tests/test_data/tracab_metadata_test.xml:50]: Test what happens when a player only has a last name (Danilo from Feyenoord): `<FirstName></FirstName>
        <LastName>Danilo</LastName>`

File: tests/test_data/tracab_td_test.dat [hunk 18]
```diff
@@ -0,0 +1,4 @@
+1509993:3,1,2,-2020,3501,0.41;1,2,34,-1350,-475,0.22;0,29,17,122,-1316,0.22;:150,-43,7,29.56,A,Alive,SetAway;:
+1509994:3,1,2,-2020,3502,0.41;1,2,34,-1350,-474,0.13;0,29,17,121,-1316,0.22;:181,-49,9,28.91,A,Alive;:
+1509995:3,1,2,-2019,3502,0.41;1,2,34,-1350,-473,0.03;0,29,17,121,-1317,0.22;:213,-56,11,29.45,A,Alive;:
+1509997:3,1,2,-2017,3503,0.41;1,2,34,-1349,-472,0.03;0,29,17,121,-1318,0.22;:276,-70,15,29.41,A,Alive;:
```
Comment (Alek050) [inline tests/test_data/tracab_td_test.dat:4]: Ball is alive and for team A the whole time, make more variations

**Review by Alek050 — CHANGES_REQUESTED** [review]

File: databallpy/load_data/tracking_data/_add_player_data_to_dict.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+    team_ids = {0: "home", 1: "away"}
+    team = team_ids.get(int(team_id))
+    if team is None:  # player is unknown or referee
+        return data
+
+    if f"{team}_{shirt_num}_x" not in data.keys():  # create keys for new player
+        data[f"{team}_{shirt_num}_x"] = [np.nan] * len(data["timestamp"])
+        data[f"{team}_{shirt_num}_y"] = [np.nan] * len(data["timestamp"])
+    #import pdb;pdb.set_trace()
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/_add_player_data_to_dict.py:26]: pdb weghalen

File: databallpy/load_data/tracking_data/_insert_missing_rows.py [hunk 20]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+import pandas as pd
+
+def _insert_missing_rows(df: pd.DataFrame, col: str) -> pd.DataFrame:
+    """Functions that inserts missing rows based on gaps in timestamp
+
+    Args:
+        df (pd.DataFrame): containing tracking data
+
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/_insert_missing_rows.py:9]: Add `col` to documentation

File: databallpy/load_data/tracking_data/tracab.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+        verbose (bool): whether to print on progress of loading in the terminal , defaults to True
+
+    Returns:
+        Tuple[pd.DataFrame, Metadata], the tracking data and metadata class
+    """
+
+    tracking_data = _get_tracking_data(tracab_loc, verbose)
+    metadata = _get_metadata(meta_data_loc)
+
+    tracking_data["match_time_td"] = [_to_match_time(x, metadata) for x in tqdm(tracking_data["timestamp"])]
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:32]: This loop is unnecessary. Maybe just input the timestamp series and the metadata and get a match time series back: `tracking_data["match_time_td"] = _to_match_time(tracking_data["timestamp"], metadata)`

File: databallpy/load_data/tracking_data/_to_match_time.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+def _to_match_time(i:int, metadata:Metadata) -> str:
+    """_summary_
+
+    Args:
+        tracking_data (pd.DataFrame): _description_
+        metadata (Metadata): _description_
+
+    Returns:
+        pd.DataFrame: _description_
+    """
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/_to_match_time.py:15]: See comment at `tracab.py`. I dont think you need to go to np.datetime64 here, takes time but we can do all the calculations in seconds. So start periods just 0, 45*60, etc. Than you can loop over the periods and apply the time for a whole period based on specific rules for that period (max/min time etc.)

File: databallpy/load_data/tracking_data/_add_player_data_to_dict.py [hunk 23]
```diff
@@ -0,0 +1,30 @@
+import numpy as np
+
+def _add_player_data_to_dict(team_id:str, shirt_num:str, x:str, y:str, data: dict, idx: int) -> dict:
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/_add_player_data_to_dict.py:3]: Maybe make clear this is tracking data data, not metadata data `_add_player_tracking_data_to_dict()`

File: databallpy/load_data/tracking_data/tracab.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+        away_score=np.nan,
+        away_formation=None,
+    )
+
+    file.close()
+
+    return metadata
+
+
+def _get_player_data(players_info: list) -> pd.DataFrame:
```
Comment (Alek050) [inline databallpy/load_data/tracking_data/tracab.py:172]: Make clear this is metadata `_get_players_metadata()`

File: tests/load_data/tracking_data/test_add_ball_data_to_dict.py [hunk 25]
```diff
@@ -0,0 +1,45 @@
+import unittest
+
+import numpy as np
+import pandas as pd
+
+from databallpy.load_data.tracking_data._add_ball_data_to_dict import _add_ball_data_to_dict
+
+
+
+class TestTracab(unittest.TestCase):
```
Comment (Alek050) [inline tests/load_data/tracking_data/test_add_ball_data_to_dict.py:10]: `class TestAddBallDataToDict`

File: tests/load_data/tracking_data/test_add_player_data_to_dict.py [hunk 26]
```diff
@@ -0,0 +1,53 @@
+import unittest
+
+import numpy as np
+
+from databallpy.load_data.tracking_data._add_player_data_to_dict import _add_player_data_to_dict
+
+
+
+class TestTracab(unittest.TestCase):
```
Comment (Alek050) [inline tests/load_data/tracking_data/test_add_player_data_to_dict.py:9]: `class TestAddPlayerDataToDict`, maybe needs to be changed if you change the file name. The format for test names is f"Test{filename}" with camel case.  The functions in the class are f"test_{function name in file you are testing}" with snake case

File: tests/load_data/tracking_data/test_insert_missing_rows.py [hunk 27]
```diff
@@ -0,0 +1,22 @@
+import unittest
+
+import pandas as pd
+
+from databallpy.load_data.tracking_data._insert_missing_rows import _insert_missing_rows
+
+class TestTracab(unittest.TestCase):
```
Comment (Alek050) [inline tests/load_data/tracking_data/test_insert_missing_rows.py:7]: Change class name

File: tests/load_data/tracking_data/test_insert_missing_rows.py [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+    def setUp(self):
+        self.input = pd.DataFrame(
+            {
+                "timestamp": [0,1,2,3,4,6,9,10,15,16,17,18,21]
+            }
+        )
+        self.expected_output = pd.DataFrame(
+            {
+                "timestamp": [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21]
```
Comment (Alek050) [inline tests/load_data/tracking_data/test_insert_missing_rows.py:16]: Maybe add one more column to check how it fills those datapoint

File: tests/load_data/tracking_data/test_tracab.py [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+        self.expected_metadata = Metadata(
+            match_id=1908,
+            pitch_dimensions=[100, 50],
+            match_start_datetime=np.datetime64("2023-01-14 13:17:00"),
+            periods_frames=pd.DataFrame(
+                {
+                    "period": [1, 2, 3, 4, 5],
+                    "start_frame": [100, 200, 300, 400, 0],
+                    "end_frame": [400, 600, 900, 1200, 0],
+                }
```
Comment (Alek050) [inline tests/load_data/tracking_data/test_tracab.py:31]: Is this changed now?

File: tests/load_data/tracking_data/test_tracab.py [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+            }
+        )
+
+    def test_get_metadata(self):
+
+        meta_data = _get_metadata(self.metadata_loc)
+        assert meta_data == self.expected_metadata
+
+    def test_get_tracking_data(self):
+        tracking_data = _get_tracking_data(self.tracking_data_loc, verbose=True)
```
Comment (Alek050) [inline tests/load_data/tracking_data/test_tracab.py:83]: For testing maybe verbose should be false

**Review by Alek050 — CHANGES_REQUESTED** [review]

File: tests/load_data/tracking_data/test_tracab.py (hunk 30 above)
Comment (DaanGro) [reply tests/load_data/tracking_data/test_tracab.py:83]: Then coverage won't be 100%

**Review by Alek050 — APPROVED** [review]
